# D38 探索性資料分析(EDA)_數據理解與重覆和遺失值處理

## 如何查詢缺失值

利用 pd.isna(df)

## 填補方式

利用 fillna 之函數達成，fillna 的 method 參數有

* bfill:由前往後補值
* ffill:由後往前補值

範例：
df['Age'] = df['Age'].fillna(df['Age'].mean())

In [2]:
import pandas as pd
import numpy as np

In [4]:
# Q1.觀察測試(test) 資料集和訓練(Train) 資料集的變數的差異性？

test = pd.read_csv("Titanic_test.csv")
train = pd.read_csv("Titanic_train.csv")

test.info()
train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 418 entries, 0 to 417
Data columns (total 11 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  418 non-null    int64  
 1   Pclass       418 non-null    int64  
 2   Name         418 non-null    object 
 3   Sex          418 non-null    object 
 4   Age          332 non-null    float64
 5   SibSp        418 non-null    int64  
 6   Parch        418 non-null    int64  
 7   Ticket       418 non-null    object 
 8   Fare         417 non-null    float64
 9   Cabin        91 non-null     object 
 10  Embarked     418 non-null    object 
dtypes: float64(2), int64(4), object(5)
memory usage: 36.0+ KB
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass  

A: train 比 test 多一個 Survived 欄位，train 有 891 筆資料，test 則有 418 筆資料。

In [11]:
# Q2.測試資料集是否有遺失值？
test

PassengerId  Pclass                                          Name  \
0            892       3                              Kelly, Mr. James   
1            893       3              Wilkes, Mrs. James (Ellen Needs)   
2            894       2                     Myles, Mr. Thomas Francis   
3            895       3                              Wirz, Mr. Albert   
4            896       3  Hirvonen, Mrs. Alexander (Helga E Lindqvist)   
..           ...     ...                                           ...   
413         1305       3                            Spector, Mr. Woolf   
414         1306       1                  Oliva y Ocana, Dona. Fermina   
415         1307       3                  Saether, Mr. Simon Sivertsen   
416         1308       3                           Ware, Mr. Frederick   
417         1309       3                      Peter, Master. Michael J   

        Sex   Age  SibSp  Parch              Ticket      Fare Cabin Embarked  
0      male  34.5      0      0              330911    7.8292   NaN        Q  
1    female  47.0      1      0              363272    7.0000   NaN        S  
2      male  62.0      0      0              240276    9.6875   NaN        Q  
3      male  27.0      0      0              315154    8.6625   NaN        S  
4    female  22.0      1      1             3101298   12.2875   NaN        S  
..      ...   ...    ...    ...                 ...       ...   ...      ...  
413    male   NaN      0      0           A.5. 3236    8.0500   NaN        S  
414  female  39.0      0      0            PC 17758  108.9000  C105        C  
415    male  38.5      0      0  SOTON/O.Q. 3101262    7.2500   NaN        S  
416    male   NaN      0      0              359309    8.0500   NaN        S  
417    male   NaN      1      1                2668   22.3583   NaN        C  

[418 rows x 11 columns]

In [20]:
test[pd.isna(test.Age)]

PassengerId  Pclass                                               Name  \
10           902       3                                   Ilieff, Mr. Ylio   
22           914       1               Flegenheim, Mrs. Alfred (Antoinette)   
29           921       3                                  Samaan, Mr. Elias   
33           925       3  Johnston, Mrs. Andrew G (Elizabeth Lily" Watson)"   
36           928       3                                Roth, Miss. Sarah A   
..           ...     ...                                                ...   
408         1300       3                    Riordan, Miss. Johanna Hannah""   
410         1302       3                             Naughton, Miss. Hannah   
413         1305       3                                 Spector, Mr. Woolf   
416         1308       3                                Ware, Mr. Frederick   
417         1309       3                           Peter, Master. Michael J   

        Sex  Age  SibSp  Parch      Ticket     Fare Cabin Embarked  
10     male  NaN      0      0      349220   7.8958   NaN        S  
22   female  NaN      0      0    PC 17598  31.6833   NaN        S  
29     male  NaN      2      0        2662  21.6792   NaN        C  
33   female  NaN      1      2  W./C. 6607  23.4500   NaN        S  
36   female  NaN      0      0      342712   8.0500   NaN        S  
..      ...  ...    ...    ...         ...      ...   ...      ...  
408  female  NaN      0      0      334915   7.7208   NaN        Q  
410  female  NaN      0      0      365237   7.7500   NaN        Q  
413    male  NaN      0      0   A.5. 3236   8.0500   NaN        S  
416    male  NaN      0      0      359309   8.0500   NaN        S  
417    male  NaN      1      1        2668  22.3583   NaN        C  

[86 rows x 11 columns]

In [28]:
# Q3.從合併資料選取一個變數，嘗試去做各種不同遺失值的處理，並透過圖形或數值來做輔助判斷，
#    補值前與後的差異，你覺得以這個變數而言，試著說明每一個方法的差異。 

test['Age'].fillna(method='ffill')

0      34.5
1      47.0
2      62.0
3      27.0
4      22.0
       ... 
413    28.0
414    39.0
415    38.5
416    38.5
417    38.5
Name: Age, Length: 418, dtype: float64

In [29]:
# 由於為由後往前，因此最後兩個值沒有辦法填補
test['Age'].fillna(method = 'bfill')

0      34.5
1      47.0
2      62.0
3      27.0
4      22.0
       ... 
413    39.0
414    39.0
415    38.5
416     NaN
417     NaN
Name: Age, Length: 418, dtype: float64